core/src/main/java/org/infinispan/util/concurrent/jdk8backported/StrippedConcurrentLinkedDeque.java

```diff
@@ -0,0 +1,723 @@
+       // trying to cas it to the last node until it does.
+       DequeNode<E> t, p, q;
+       restartFromTail:
+       while ((t = tail).item == null && (p = t.next) != null) {
+           for (;;) {
+               if ((q = p.next) == null ||
+                   (q = (p = q).next) == null) {
+                   // It is possible that p is NEXT_TERMINATOR,
+                   // but if so, the CAS is guaranteed to fail.
+                   if (casTail(t, p))
+                       return;
+                   else
+                       continue restartFromTail;
+               }
+               else if (t != tail)
+                   continue restartFromTail;
+               else
+                   p = q;
+           }
+       }
+   }
+
+   void skipDeletedPredecessors(DequeNode<E> x) {
+       whileActive:
+       do {
+           DequeNode<E> prev = x.prev;
+           // assert prev != null;
+           // assert x != NEXT_TERMINATOR;
+           // assert x != PREV_TERMINATOR;
+           DequeNode<E> p = prev;
+           findActive:
+           for (;;) {
+               if (p.item != null)
+                   break findActive;
+               DequeNode<E> q = p.prev;
+               if (q == null) {
+                   if (p.next == p)
+                       continue whileActive;
+                   break findActive;
+               }
+               else if (p == q)
+                   continue whileActive;
+               else
+                   p = q;
+           }
+
+           // found active CAS target
+           if (prev == p || x.casPrev(prev, p))
+               return;
+
+       } while (x.item != null || x.next == null);
+   }
+
+   void skipDeletedSuccessors(DequeNode<E> x) {
+       whileActive:
+       do {
+           DequeNode<E> next = x.next;
+           // assert next != null;
+           // assert x != NEXT_TERMINATOR;
+           // assert x != PREV_TERMINATOR;
+           DequeNode<E> p = next;
+           findActive:
+           for (;;) {
+               if (p.item != null)
+                   break findActive;
+               DequeNode<E> q = p.next;
+               if (q == null) {
+                   if (p.prev == p)
+                       continue whileActive;
+                   break findActive;
+               }
+               else if (p == q)
+                   continue whileActive;
+               else
+                   p = q;
+           }
+
+           // found active CAS target
+           if (next == p || x.casNext(next, p))
+               return;
+
+       } while (x.item != null || x.prev == null);
+   }
+
+   /**
+    * Returns the predecessor of p, or the last node if p.prev has been
+    * linked to self, which will only be true if traversing with a
+    * stale pointer that is now off the list.
+    */
+   final DequeNode<E> pred(DequeNode<E> p) {
+       DequeNode<E> q = p.prev;
+       return (p == q) ? last() : q;
+   }
+
+   public DequeNode<E> peekFirstNode() {
```

Comment:
The atomicity was in the usage of the command itself.  Unfortunately since it is just a peek by the time it returns the value may not be the first or last anymore so I couldn't rely on it :(